### PR TITLE
Feature/upload field

### PR DESCRIPTION
**Pull request by ibelar**
## Implement Upload and UploadImg field.

Field instance must use two callbacks, onUpload and onDelete.

```
$field = $form->addField('file', ['Upload', ['accept' => ['.png', '.jpg']]]);
$img = $form->addField('file', ['UploadImg']);
```
- UploadImg extends Upload. It add a preview thumbnail view next to the field. 

You can then set thumbnail path from the upload callback using:
```
$img->setThumbnailSrc('./images/test.jpg');
```
You may also clear thumbnail path or supply a default one using UploadImg::clearThumbnail()

### onUpload callback

Call after user select a file in browser dialog. Callback will either pass the uploaded file, i.e. $_FILES, or 'error' as parameter. It is up to developper to check for error and properly store file on the server.

```
$field->onUpload(function ($files) use ($form, $field) {
    if ($files === 'error') {
        return $form->error('file', 'Error uploading file.');
    }
    //Do file processing here...

   /* This will get caught by jsCallback and show via modal. */
   //new Blabla();

   /* js Action can be return. */
   //if using form, can return an error to form field directly.
   //return $form->error('file', 'Unable to upload file.');

   // can also return a notifier.
   return new atk4\ui\jsNotify(['content' => 'File is uploaded!', 'color' => 'green']);
});
```
The callback may return js actions object to display error content or simply notify the user that the file has been successfully store on server.

### onDelete callback

Call after a file has been upload and user simply want to remove it by clicking the remove button. The callback will receive the filename as parameter.

```
$field->onDelete(function ($fileName) {
    return new atk4\ui\jsNotify(['content' => $fileName.' has been removed!','color' => 'green']);
});
```
Again, you can return a js action to display status to user.

Note: You will need to update atk4JS in order to run this PR.

### See upload.php in demos for more info.

### CODE REVIEW

**Comment by ibelar** [comment]
Refactor some of the js file in my last commit.
This way, all atk services are exposed to the outside world using the global var atk.

For example, another script from an add-on will now be able to upload files via the uploadService by using  atk.uploadService.uploadFile() method. 
Same for api service.  If someone want to create their own sse event plugin, they can now use atk.apiService.atkSuccessTest(jsonResponse) globally.

File: src/FormField/Upload.php [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+    /**
+     * Allow multiple file or not.
+     * CURRENTLY NOT SUPPORTED.
+     *
+     * @var bool
+     */
+    public $multiple = false;
+
+    /**
+     * An array of string value for accept file type.
```
[the remark this answers is not in the document]
Comment (ibelar) [reply src/FormField/Upload.php]: Yes will prevent from choosing multiple file in user dialog.
Will check on multiple upload later.

File: src/FormField/Upload.php [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+     * @var array
+     */
+    public $accept = [];
+
+    /**
+     * The.
+     *
+     * @var null
+     */
+    //public $fieldIdName = null;
```
[the remark this answers is not in the document]
Comment (ibelar) [reply src/FormField/Upload.php]: was changed to fiIeId...

File: src/FormField/Upload.php [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+    {
+        $this->fileId = $id;
+    }
+
+    /**
+     * Add a js action to be return to server on callback.
+     *
+     * @param $action
+     */
+    public function addJsAction($action)
```
[the remark this answers is not in the document]
Comment (ibelar) [reply src/FormField/Upload.php]: ok fixed

File: js/src/plugins/fileUpload.js [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+   */
+  doFileDelete(fileName) {
+
+    const that = this;
+
+    this.$el.api({
+      on: 'now',
+      url: this.settings.uri,
+      data: {'action': 'delete', 'f_name': fileName},
+      method: 'POST',
```
[the remark this answers is not in the document]
Comment (ibelar) [reply js/src/plugins/fileUpload.js]: will update this in new js package if merge.

File: src/FormField/Upload.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (romaninsh) [reply src/FormField/Upload.php]: hi i don't see any new commits. Did you push? Also - "git pull origin/master"

**Comment by romaninsh** [comment]
works as advertised, although lacks documentation.

**Pull request merged**